File: python/ine/ine_jaxi_function.py
import io
from typing import Optional

import requests
import pandas as pd


def ine_jaxi_api_function(
    table_id: str,
    nocab: str = "1",
    directory: str = "t",
    locale: str = "es",
    variant: str = "csv_bdsc",
) -> pd.DataFrame:
    """
    Descarga un archivo CSV de INE JAXIT3 y retorna un pandas DataFrame.

    Parámetros
    ----------
    table_id : str
        Identificador de la tabla INE (p. ej., "67821").
    nocab : str, optional
        Control de cabecera; "1" evita cabeceras adicionales. Por defecto "1".
    directory : str, optional
        Segmento de directorio. Por defecto "t".
    locale : str, optional
        Idioma del recurso. Por defecto "es".
    variant : str, optional
        Variante del CSV. Por defecto "csv_bdsc".

    Retorna
    -------
    pandas.DataFrame
        CSV parseado como un DataFrame con columnas de texto.
    """

    # URL de la API de INE JAXIT3
    base_url = "https://www.ine.es/jaxiT3/files"
    url = f"{base_url}/{directory}/{locale}/{variant}/{table_id}.csv"

    # Parámetros de la petición
    headers = {"Accept": "text/csv"}
    params = {"nocab": nocab}

    # Hacemos la petición
    resp = requests.get(url, headers=headers, params=params, timeout=60)
    try:
        resp.raise_for_status()
    except requests.HTTPError as exc:
        raise RuntimeError(f"INE JAXIT3 request failed [{resp.status_code}]") from exc

    # Parseamos el CSV con pandas
    text_stream = io.StringIO(resp.content.decode("utf-8-sig"))
    df: pd.DataFrame = pd.read_csv(text_stream, sep=";", dtype=str)
    return df

File: R/ine/ine_jaxi_function.R
# Función mínima para descargar datos INE JAXIT3 (CSV)

# Inputs obligatorios:
# - tableId: identificador de la tabla INE (p.ej., "67821")
#
# Inputs opcionales:
# - nocab: controla cabecera ("1" evita cabeceras adicionales)
# - directory: segmento de directorio (por defecto "t")
# - locale: idioma (por defecto "es")
# - variant: variante de CSV (por defecto "csv_bdsc")
#
# Output:
# - df: data.frame con los datos descargados

ine_jaxi_api_function <- function(
  tableId,
  nocab = "1",
  directory = "t",
  locale = "es",
  variant = "csv_bdsc"
) {
  base_url <- "https://www.ine.es/jaxiT3/files"
  url <- paste0(base_url, "/", directory, "/", locale, "/", variant, "/", tableId, ".csv")

  res <- httr::GET(url, query = list(nocab = nocab), httr::accept("text/csv"))
  if (httr::http_error(res)) stop(sprintf("INE JAXIT3 request failed [%s]", httr::status_code(res)))

  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  df <- utils::read.csv(text = txt, stringsAsFactors = FALSE, sep = ";")
  return(df)
}


